File: src/applicationruntime.cpp
// SPDX-FileCopyrightText: 2026 KIM Hyunjae
// SPDX-License-Identifier: AGPL-3.0-or-later

#include "applicationruntime.h"

#include "localization.h"

#include <QGuiApplication>
#include <QQmlApplicationEngine>
#include <QQuickStyle>
#include <QString>
#include <QUrl>
#include <QVariant>
#include <QtGlobal>

namespace {
void setupApplicationIdentity()
{
    QGuiApplication::setDesktopFileName(QStringLiteral("io.github.hnjae.KiriView"));
}

void setupDefaultQuickStyle()
{
    if (!qEnvironmentVariableIsSet("QT_QUICK_CONTROLS_STYLE")) {
        QQuickStyle::setStyle(QStringLiteral("org.kde.desktop"));
    }
}

QUrl validInitialSourceUrl(const QUrl &url)
{
    if (url.isEmpty() || !url.isValid()) {
        return QUrl();
    }

    return url;
}
}

namespace KiriView {
void initializeApplicationRuntime()
{
    initializeLocalization();
    setupApplicationIdentity();
    setupDefaultQuickStyle();
}

QUrl initialSourceUrlFromLocalFilePath(const QString &path)
{
    return validInitialSourceUrl(QUrl::fromLocalFile(path));
}

QUrl initialSourceUrlFromUrlText(const QString &urlText)
{
    return validInitialSourceUrl(QUrl(urlText));
}

void loadApplicationMainQml(QQmlApplicationEngine &engine, const QUrl &initialSourceUrl)
{
    setupLocalizedContext(engine);

    if (!initialSourceUrl.isEmpty()) {
        QVariantMap initialProperties;
        initialProperties.insert(QStringLiteral("initialSourceUrl"), initialSourceUrl);
        engine.setInitialProperties(initialProperties);
    }

    engine.load(QUrl(QStringLiteral("qrc:/qt/qml/io/github/hnjae/kiriview/src/qml/Main.qml")));
}
}

File: src/applicationruntime.rs
// SPDX-FileCopyrightText: 2026 KIM Hyunjae
// SPDX-License-Identifier: AGPL-3.0-or-later

use crate::startup_arguments::StartupSource;
use cxx_qt_lib::QQmlApplicationEngine;
use std::pin::Pin;

#[cxx::bridge(namespace = "KiriView")]
mod ffi {
    enum ApplicationStartupSourceKind {
        None,
        LocalFilePath,
        UrlText,
    }

    struct ApplicationStartupSource {
        kind: ApplicationStartupSourceKind,
        text: String,
    }

    unsafe extern "C++" {
        include!("cxx-qt-lib/qqmlapplicationengine.h");
        #[namespace = ""]
        type QQmlApplicationEngine = cxx_qt_lib::QQmlApplicationEngine;

        include!("applicationruntime.h");

        #[cxx_name = "initializeApplicationRuntime"]
        fn initialize_application_runtime();

        #[cxx_name = "loadApplicationMainQml"]
        fn load_application_main_qml(
            engine: Pin<&mut QQmlApplicationEngine>,
            startup_source: &ApplicationStartupSource,
        );
    }
}

pub use ffi::{ApplicationStartupSource, ApplicationStartupSourceKind};

pub fn initialize_application_runtime() {
    ffi::initialize_application_runtime();
}

pub fn application_startup_source(source: Option<StartupSource>) -> ApplicationStartupSource {
    match source {
        Some(StartupSource::LocalFile(path)) => ApplicationStartupSource {
            kind: ApplicationStartupSourceKind::LocalFilePath,
            text: path.to_string_lossy().into_owned(),
        },
        Some(StartupSource::Url(url)) => ApplicationStartupSource {
            kind: ApplicationStartupSourceKind::UrlText,
            text: url,
        },
        None => ApplicationStartupSource {
            kind: ApplicationStartupSourceKind::None,
            text: String::new(),
        },
    }
}

pub fn load_application_main_qml(
    engine: Pin<&mut QQmlApplicationEngine>,
    startup_source: &ApplicationStartupSource,
) {
    ffi::load_application_main_qml(engine, startup_source);
}
